packages/astro/src/core/build/plugins/plugin-css.ts

```diff
@@ -25,236 +25,312 @@ interface PluginOptions {
-			enforce: 'post',
-			configResolved(config) {
-				resolvedConfig = config;
-			},
-			generateBundle(_, bundle) {
-				// If user disable css code-splitting, search for Vite's hardcoded
-				// `style.css` and add it as css for each page.
-				// Ref: https://github.com/vitejs/vite/blob/b2c0ee04d4db4a0ef5a084c50f49782c5f88587c/packages/vite/src/node/plugins/html.ts#L690-L705
-				if (!resolvedConfig.build.cssCodeSplit) {
-					const cssChunk = Object.values(bundle).find(
-						(chunk) => chunk.type === 'asset' && chunk.name === 'style.css'
-					);
-					if (cssChunk) {
-						for (const pageData of eachPageData(internals)) {
-							pageData.css.set(cssChunk.fileName, { depth: -1, order: -1 });
-						}
-					}
-				}
-			},
+	};
+
+	const singleCssPlugin: VitePlugin = {
+		name: 'astro:rollup-plugin-single-css',
+		enforce: 'post',
+		configResolved(config) {
+			resolvedConfig = config;
 		},
-	];
-}
+		generateBundle(_, bundle) {
+			// If user disable css code-splitting, search for Vite's hardcoded
+			// `style.css` and add it as css for each page.
+			// Ref: https://github.com/vitejs/vite/blob/b2c0ee04d4db4a0ef5a084c50f49782c5f88587c/packages/vite/src/node/plugins/html.ts#L690-L705
+			if (resolvedConfig.build.cssCodeSplit) return;
+			const cssChunk = Object.values(bundle).find(
+				(chunk) => chunk.type === 'asset' && chunk.name === 'style.css'
+			);
+			if (cssChunk === undefined) return;
+			for (const pageData of eachPageData(internals)) {
+				const cssToInfoMap =
+					pagesToCss.get(pageData.moduleSpecifier) ??
+					pagesToCss.set(pageData.moduleSpecifier, new Map()).get(pageData.moduleSpecifier)!;
 
-export function pluginCSS(
-	options: StaticBuildOptions,
-	internals: BuildInternals
-): AstroBuildPlugin {
-	return {
-		build: 'both',
-		hooks: {
-			'build:before': ({ build }) => {
-				let plugins = rollupPluginAstroBuildCSS({
-					buildOptions: options,
-					internals,
-					target: build === 'ssr' ? 'server' : 'client',
-				});
+				cssToInfoMap.set(cssChunk.fileName, { depth: -1, order: -1 });
+			}
+		},
+	};
 
-				return {
-					vitePlugin: plugins,
-				};
-			},
+	const inlineStylesheetsPlugin: VitePlugin = {
+		name: 'astro:rollup-plugin-inline-stylesheets',
+		enforce: 'post',
+		async generateBundle(_outputOptions, bundle) {
+			const inlineConfig = settings.config.experimental.inlineStylesheets;
+			const { assetsInlineLimit = 4096 } = settings.config.vite?.build ?? {};
+
+			Object.entries(bundle).forEach(([id, stylesheet]) => {
+				if (
+					stylesheet.type !== 'asset' ||
+					stylesheet.name?.endsWith('.css') !== true ||
+					typeof stylesheet.source !== 'string'
+				)
+					return;
+
+				const assetSize = new TextEncoder().encode(stylesheet.source).byteLength;
+
+				const toBeInlined =
+					inlineConfig === 'always'
+						? true
+						: inlineConfig === 'never'
+						? false
+						: assetSize <= assetsInlineLimit;
+
+				if (toBeInlined) delete bundle[id];
```

Comment:
Unless I've missed something, this _could_ delete CSS chunks that are only used by framework components. I have not tested if it's true though.

Reply:
Why do you think that? A chunk used by a framework component is still injected into the page.

Reply:
I just check it locally and this doesn't seem to be an issue, so ignore me 😄 I thought that CSS chunk could be not linked to any page, but seems like it'll always be.